### PR TITLE
Add support for 1-second Storage Resolution in the AWS EMF Exporter

### LINKED ISSUE #29506: [exporter/awsemf] Add StorageResolution config to support high-resolution metrics
### Component(s)

exporter/awsemf

### Is your feature request related to a problem? Please describe.

We want to be able to push high-resolution metrics from the opentelemetry-collector. This means that we should be able to set StorageResolution to 1.

### Describe the solution you'd like

Have the option in the AWSEMF config to turn on high-resolution metrics through the StorageResolution config.

### Describe alternatives you've considered

There doesn't seem to be another way of doing this, except for creating a new exporter that uses either this setting, or directly uses [PutMetricData](https://docs.aws.amazon.com/AmazonCloudWatch/latest/APIReference/API_PutMetricData.html), which would mean a completely new exporter.

### Additional context

I will be creating a PR soon.

### PR DESCRIPTION
**Pull request by jpbarto**
This change implements the capability for users of the AWS EMF Exporter to specify which metrics they would like to have sent to CloudWatch with a 1 second Storage Resolution. The EMF Exporter now explicitly states the Storage Resolution for each metric as 60 seconds, the previous implicit default, so there is no behavior change. If the user specifies a metric to have 1 second resolution it will be sent to CloudWatch EMF with the Storage Resolution set accordingly.

#### Description
Previously the AWS EMF Exporter sent metric data into CloudWatch without specifying the storage resolution. CloudWatch would then default to a 60 second storage resolution, even if metrics are sent more frequently than every 60 seconds. This would confuse users when they try to apply functions like AVG, SUM, MAX, or MIN to their metrics with a period of 5 seconds. The function would be applied by CloudWatch to 60 seconds worth of data and produced unexpected results and confusion for the user. This commit makes this 60 second resolution explicit in the messages sent to CloudWatch by the EMF Exporter and also gives the user the option to specify a more granular 1 second resolution per metric in the configuration file of the AWS EMF Exporter.

#### Link to tracking issue
Fixes #29506 

#### Testing
Added tests to verify that config file parsing validates a metric descriptor that specifies either a valid unit, valid storage resolution, or both and rejects other invalid metric descriptors.
Added tests that the translation from metric data to CW EMF carries a storage resolution with it, defaulting to a value of 60 (current behavior) if no storage resolution valid is explicitly set in the configuration.

#### Documentation
Comments added in the code but have not updated the README.md pending acceptance of the PR.

### CODE REVIEW

File: exporter/awsemfexporter/config.go
Comment (Aneurysm9) [inline exporter/awsemfexporter/config.go:105]: The `MetricDescriptor` configuration is extremely limited and can be accomplished through generally-applicable processors.  I think it should be deprecated and slated for removal at some point in the future so we should not expand its surface area any further.

File: exporter/awsemfexporter/metric_translator.go
Comment (Aneurysm9) [inline exporter/awsemfexporter/metric_translator.go:51]: I don't like the use of `any` here.  Use the type system to prevent invalid data from being represented.  If that means a new type needs to be created then a new type will need to be created.  Rather than `[]map[string]any` this should be something like `[]metricDefinition` with `metricDefinition` defined like this:

```go
type metricDefinition struct {
  Name string
  Unit string
  StorageResolution int
}
```

Comment (jpbarto) [reply exporter/awsemfexporter/metric_translator.go]: The latest update creates a `cWMetricInfo` struct in the translator to mimic the `metricInfo` struct in grouped_metric.go, replacing the use of `map[string]any` in the translator.

File: exporter/awsemfexporter/config.go
Comment (jpbarto) [reply exporter/awsemfexporter/config.go]: The latest update instead looks for a metric attribute named `aws.emf.storage_resolution` which can be applied either in the application's code or using the Attribute Processor in the collector itself.

**Comment by jpbarto** [comment]
@Aneurysm9 are you able to approve the change in its current form which now uses metric attributes to define storage resolution?

File: exporter/awsemfexporter/metric_translator.go
Comment (mxiamxia) [inline exporter/awsemfexporter/metric_translator.go:226]: Please remove all the debug logs

File: exporter/awsemfexporter/metric_translator.go
Comment (mxiamxia) [inline exporter/awsemfexporter/metric_translator.go:511]: I'm concerned about the performance impact of applying regex matching on every metric attribute for each grouped metric data point. 
Is there any requirement to add more `aws.emf.*` attribute patterns? Could you perform an exact match for `labelname == "aws.emf.storage_resolution"` and break the loop once it's found?

File: exporter/awsemfexporter/metric_translator.go
Comment (mxiamxia) [inline exporter/awsemfexporter/metric_translator.go:229]: ditto

File: exporter/awsemfexporter/metric_translator.go
Comment (mxiamxia) [inline exporter/awsemfexporter/metric_translator.go:249]: ditto

File: exporter/awsemfexporter/metric_translator.go
Comment (mxiamxia) [inline exporter/awsemfexporter/metric_translator.go:37]: q: why is regex rule is `^awsemf.[A-Za-z_]*$` instead of `^aws.emf.[A-Za-z_]*$`?

Comment (jpbarto) [reply exporter/awsemfexporter/metric_translator.go]: Thank you @mxiamxia - I've swapped the more complex regex for a simpler comparison operator. As new metric attributes are added this can be expanded to a switch statement which will likely still perform better than a compiled regex.

**Review by mxiamxia — APPROVED** [review]
LGTM. ty!

**Comment by jpbarto** [comment]
@andrzej-stencel good morning, are you able to review this pull request please?

**Review by andrzej-stencel — APPROVED** [review]

**Comment by andrzej-stencel** [comment]
@jpbarto this probably deserves a changelog entry? Can you add one please https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/CONTRIBUTING.md#changelog.

**Comment by jpbarto** [comment]
@andrzej-stencel the changelog has been updated as requested.

**Comment by jpbarto** [comment]
@andrzej-stencel @mxiamxia the PR requires a second reviewer to approve the pull request. This is currently assigned to @bryan-aguilar. Bryan was involved with the original issue raised last year and has been idle on GitHub since June of this year. Can I please ask that this PR be assigned to an alternate reviewer?

**Comment by jpbarto** [comment]
@andrzej-stencel this PR has been approved by both yourself and one of the code owners, Min Xia (see above). Please could you mark this PR as `ready to merge`?

**Pull request merged**
